Optional args parsing

maringuu opened the pull request:
I've implemented your suggestion from https://github.com/ifreund/river/issues/102#issuecomment-693645836.

There are several things left to be discussed and solved.
I've implemented options with an Int argument too to figure out a good way to parse the argument.
When merging this I will of course use this for `map -release` and remove the Int option as it is not needed.

### Default Options for bools
Since boolean options are either given or not given I'd like to make the default be false always and change them to true when the argument is given. 
A possibility would be to make it a compile error to have the default value set to false. This is what I implemented currently as it seems clean and explicit about the default beeing false.
Another idea would be to make giving the option toggle the default value. Code wise this would be the simplest but I really dislike this one as there is really no use for that imo.

### Enforce naming of arguments
Since we decided to always have arguments be in the form `-long-option` we could implement a check to enforce correct naming at compile time. Is there anything bad about this?

### Setting the offset
The last thing I'm not happy with is setting the offset. The offset is needed for the caller to know where to continue parsing.
If there was a way to run an else branch after a for loop even when the for loop breaks it would be nice to set it there.
There was a issue in the zig bugtracker somewhere but I can't seem to find it.

Closes #102

maringuu commented:
The newest commit somehow does not compile. I don't see why atm.

ifreund commented on river/command.zig:
```diff
@@ -117,3 +117,71 @@ pub fn errToMsg(err: Error) [:0]const u8 {
+    defer offset.* = i;
+
+    outer: for (args) |arg, idx| {
+        i = idx + 1;
```
Pretty sure this is the why the tests are crashing, should just be `i = idx`

maringuu replied:
I changed it to just count the parsed arguments by adding one each time.

ifreund commented on river/command.zig:
```diff
@@ -117,3 +117,71 @@ pub fn errToMsg(err: Error) [:0]const u8 {
+                    .Int => {
+                        if (args.len < idx + 2) return error.MissingArgument;
+                        @field(ret, field.name) = try std.fmt.parseInt(field.field_type, args[idx+1], 10);
+                        continue :outer;
```
This is insufficient as the next iteration must be skipped as well, but it doesn't matter as we don't need to parse ints yet :)

maringuu replied:
Yeah. Removed it.

ifreund reviewed (commented):
> ...and remove the Int option as it is not needed.

+1, I'm starting to question if there will be ever be a need for something beyond bools tbh.

> Since boolean options are either given or not given I'd like to make the default be false always and change them to true when the argument is given.

This sounds reasonable to me.

maringuu commented:
I removed the int parsing now and changed `map.zig` to use this parsing.
In `command.zig` line 136 there is a todo left which is about enforcing the names of parameters to match a specific format.

ifreund commented on river/command.zig:
```diff
@@ -117,3 +117,59 @@ pub fn errToMsg(err: Error) [:0]const u8 {
+            @compileError("option names must start with a single - followed by an alphabetical cahr");
+    }
+
+    // Counts the amout of parsed args
+    var i: usize = 0;
+    // Do the actual parsing of args
+    for (args) |arg| {
+        var found = false;
+        inline for (@typeInfo(T).Struct.fields) |field| {
+            if (std.mem.eql(u8, arg, field.name)) {
+                found = true;
+                @field(ret, field.name) = true;
+            }
+        }
+        // stop parsing options when no options are left
+        if (!found) break;
+        i += 1;
+    }
+
+    offset.* = i;
```
This loop can be written a little more cleanly:
```suggestion
    offset.* = 0;
    for (args) |arg| {
        inline for (@typeInfo(T).Struct.fields) |field| {
            if (std.mem.eql(u8, arg, field.name)) {
                @field(ret, field.name) = true;
                offset.* += 1;
                continue;
            }
        }
        // There are no options left
        break;
    }
```

ifreund commented on river/command.zig:
```diff
@@ -117,3 +117,59 @@ pub fn errToMsg(err: Error) [:0]const u8 {
+            @compileError("unsupported option type " ++ field.field_type);
+        if (field.default_value.?)
+            @compileError("bool options must be false by default");
+        if (field.name[0] != '-')
+            // TODO Should we really have this check? If yes test ascii.isAlpha(field.name[1])
+            // Figure out how long arguments should be is -a a valid argument
+            @compileError("option names must start with a single - followed by an alphabetical cahr");
```
I'd say just drop this check. Options not conforming to the naming convention will be fairly easy to catch in review.

ifreund commented on river/command/map.zig:
```diff
@@ -74,7 +77,9 @@ pub fn map(
-        if (existant_mapping.modifiers == modifiers and existant_mapping.keysym == keysym and existant_mapping.release == optionals.release) {
+        if (existant_mapping.modifiers == modifiers
+            and existant_mapping.keysym == keysym
+            and existant_mapping.release == @field(options, "-release")) {
```
Pretty sure you can just do this here:
```suggestion
            and existant_mapping.release == options.@"-release") {
```
`@field()` is only needed if the name isn't static at comptime.

maringuu commented:
Sorry I was busy and didn't have time for this pr.
It should be finished now.

maringuu commented:
Oops i accidently closed it

maringuu commented:
I dont know how I keep closing this over and over again.

maringuu commented:
Rebased.

maringuu commented:
Rebased again. It is now used for `unmap` too.

maringuu commented:
This in neat and all but I've come to the conclusion that we dont yet need a fancy way to do parsing of arguments.
